### PR TITLE
Add buildDirs to build method

### PR DESCRIPTION
**Pull request by grouma**
- Update `run` method to take in `buildDirs`
- Remove `buildDirs` from `BuildOptions`
- Rev package as this is a breaking change

### CODE REVIEW

File: _test_common/lib/test_phases.dart [hunk 1]
```diff
@@ -129,7 +129,7 @@ Future<BuildResult> testBuilders(
     builderConfigOverrides,
     isReleaseBuild: false,
   );
-  result = await build.run({});
+  result = await build.run({}, options.buildDirs);
```
[the remark this answers is not in the document]
Comment (grouma) [reply _test_common/lib/test_phases.dart]: The `OptionalOutputTracker` requires a list of buildDirs. This becomes somewhat problematic due to the order of construction.

[the remark this answers is not in the document]
Comment (natebosch) [reply _test_common/lib/test_phases.dart]: I think that probably means that the `OptionalOutputTracker` won't work when we need it for the build daemon?

File: build_runner_core/lib/src/generate/build_impl.dart [hunk 2]
```diff
@@ -85,11 +84,12 @@ class BuildImpl {
         _verbose = options.verbose,
         _environment = buildDefinition.environment,
         _trackPerformance = options.trackPerformance,
-        _buildDirs = options.buildDirs,
         _logPerformanceDir = options.logPerformanceDir;
 
-  Future<BuildResult> run(Map<AssetId, ChangeType> updates) =>
-      _SingleBuild(this).run(updates)..whenComplete(_resolvers.reset);
+  Future<BuildResult> run(
+          Map<AssetId, ChangeType> updates, List<String> buildDirs) =>
```
Comment (jakemac53) [inline build_runner_core/lib/src/generate/build_impl.dart]: Why not make `buildDirs` an optional (named) argument instead (with [] as the default)? Then i don't think we have to deal with a breaking change (and its a sane default anyways).

Comment (natebosch) [reply build_runner_core/lib/src/generate/build_impl.dart]: I think it would still be a breaking behavior change

Comment (grouma) [reply build_runner_core/lib/src/generate/build_impl.dart]: This is still a breaking change regardless as we expose `BuildOptions`. I do like making the buildDirs optional though.

**Pull request merged**